File: Graph/bfs.cpp
#include <bits/stdc++.h>
using namespace std;
map<int,vector<int>>adjlist;
map<int,bool>visited;
map<int,int>dstance;
map<int,int>parent;

void bfs(int startingNode)
{
    queue<int>q;
    q.push(startingNode);
    dstance[startingNode] = 0;
    visited[startingNode] = true;

    while (!q.empty())
    {
        int x = q.front();
        cout<<"x "<<x<<endl;
        q.pop();
        for (int i =0;i < adjlist[x].size();i++)
        {
            if(visited[adjlist[x][i]] == false)
            {
                dstance[adjlist[x][i]] = dstance[x]+1;
                visited[adjlist[x][i]] = true;
                q.push(adjlist[x][i]);
                parent[adjlist[x][i]] = x;
            }
        }
        cout<<endl;
    }
}

int main()
{
    int node,edge;
    cin>>node>>edge;
    
    while(edge--)
    {
        int x,y; cin>>x>>y;
        adjlist[x].push_back(y);
        adjlist[y].push_back(x);
    }
    int startingNode = 1;
    bfs(startingNode);
    cout<<"shortest path from node "<<startingNode<<" to 10 is = "<<dstance[10]<<endl;
}

/*
10 13
1 2
1 3
1 4
2 6
3 7
3 8
4 7
5 10
6 10
7 8
7 9
8 5
9 10

this sample input's graph diagram>  http://www.shafaetsplanet.com/planetcoding/wp-content/uploads/2014/02/bfs4-300x296.png
*/

File: Graph/bfs.py
from collections import deque,defaultdict
def bfs(startingNode,adjlist,visited,parent):
    visited[startingNode] = True
    distance[startingNode] = 0
    q = deque()
    q.appendleft(startingNode)
    
    while q:
        x = q.pop()
        for i in adjlist[x]:
            if not visited[i]:
                q.appendleft(i)
                visited[i] = True
                parent[i] = x
                distance[i] = distance[x] + 1
    #         elif parent[x] != i:  # Cycle detection
    #             return True
    # return False
    
if __name__ == '__main__':
    node,edges = map(int,input().split())
    # creating list
    # adjlist = {i:[] for i in range(1,node+1)}   # with normal dict
    # visited = [False]*(node+1)
    # parent = [0]*(node+1)
    # distance = [0]*(node+1)
    
    # maping key by default dic to handle larger node
    adjlist = defaultdict(list) 
    visited = defaultdict(bool)
    parent = defaultdict(int)
    distance = defaultdict(int)
    
    for _ in range(edges):
        x,y = map(int,input().split())
        adjlist[x].append(y)
        adjlist[y].append(x)
        
    startingNode = 1
    bfs(startingNode,adjlist,visited,parent)
    endnode = 10
    print("Distance from 1 to",endnode,"is =",distance[endnode])

"""    Sample input
10 13
1 2
1 3
1 4
2 6
3 7
3 8
4 7
5 10
6 10
7 8
7 9
8 5
9 10

"""